src/hipscat_import/catalog/margin_cache.py

```diff
@@ -0,0 +1,249 @@
+import healpy as hp
+import numpy as np
+import pandas as pd
+from dask.distributed import Client, as_completed
+from hipscat import pixel_math
+from hipscat.io import file_io, paths
+from tqdm import tqdm
+
+# pylint: disable=too-many-locals,too-many-arguments
+
+def _find_parition_margin_pixel_pairs(stats, margin_order):
+    """Creates a DataFrame filled with many-to-many connections between
+    the catalog partition pixels and the margin pixels at `margin_order`.
+    """
+    norders = []
+    part_pix = []
+    margin_pix = []
+
+    for _, row in stats.iterrows():
+        order = row["Norder"]
+        pix = row["Npix"]
+
+        d_order = margin_order - order
+
+        margins = pixel_math.get_margin(order, pix, d_order)
+
+        for m_p in margins:
+            norders.append(order)
+            part_pix.append(pix)
+            margin_pix.append(m_p)
+
+    margin_pairs_df = pd.DataFrame(
+        zip(norders, part_pix, margin_pix),
+        columns=["partition_order", "partition_pixel", "margin_pixel"]
+    )
+    return margin_pairs_df
+
+def _create_margin_directory(stats, output_path):
+    """Creates directories for all the catalog partitions."""
+    for _, row in stats.iterrows():
+        order = row["Norder"]
+        pix = row["Npix"]
+
+        destination_dir = paths.pixel_directory(
+            output_path, order, pix
+        )
+        file_io.make_directory(destination_dir, exist_ok=True)
+
+def _map_to_margin_shards(client, args, partition_pixels, margin_pairs):
+    """Create all the jobs for mapping partition files into the margin cache."""
+    futures = []
+    for pix in partition_pixels:
+        partition_file = paths.pixel_catalog_file(
+            args.input_catalog_path,
+            pix.order,
+            pix.pixel
+        )
+        futures.append(
+            client.submit(
+                _map_pixel_shards,
+                partition_file=partition_file,
+                margin_pairs=margin_pairs,
+                margin_threshold=args.margin_threshold,
+                output_path=args.margin_output_path,
+                margin_order=args.margin_order,
+                ra_column=args.catalog.catalog_info.ra_column,
+                dec_column=args.catalog.catalog_info.dec_column,
+            )
+        )
+
+    for _ in tqdm(
+        as_completed(futures, with_results=True),
```

Comment:
Will the mapping stage have a return value?

Reply:
I don't have one set up for it atm

Reply:
is there something I should be returning? 😗